test/test_integration.rb

```diff
@@ -389,7 +389,207 @@ def test_no_zombie_children
     worker_pids.each { |pid| Process.kill :TERM, pid }
     sleep 2
 
+    # should be empty if all old workers removed
+    waited_pids = worker_pids.map { |pid|
+      begin
+        Process.wait pid, Process::WNOHANG
+        pid
+      rescue Errno::ECHILD
+        nil # child is already reaped/removed
+      end
+    }.compact
+
     # Should return nil if Puma has correctly cleaned up
-    assert_nil Process.waitpid(-1, Process::WNOHANG)
+    assert_empty waited_pids
+  end
+
+  def test_worker_spawn_external_term
+    skip NO_FORK_MSG unless HAS_FORK
+
+    host = '127.0.0.1'
+    port = UniquePort.call
+    workers_booted = 0
+
+    conf = Puma::Configuration.new do |c|
+      c.bind "tcp://#{host}:#{port}"
+      c.threads 1, 1
+      c.workers 2
+      c.worker_shutdown_timeout 2
+      c.app TestApps::SLEEP
+      c.after_worker_fork { |idx| workers_booted += 1 }
+    end
+
+    l = Puma::Launcher.new conf, :events => @events
+
+    t = Thread.new do
+      Thread.current.abort_on_exception = true
+      l.run
+    end
+
+    wait_booted
+
+    # make sure two workers have booted
+    time = 0
+    until workers_booted >= 2 || time >= 10
+      sleep 2
+      time += 2
+    end
+
+    cluster = l.instance_variable_get(:@runner)
+
+    http0 = Net::HTTP.new host, port
+    http1 = Net::HTTP.new host, port
+
+    worker0 = Thread.new do
+      begin
+        req0 = Net::HTTP::Get.new '/sleep35'
+        http0.start.request(req0) { |rep0| rep0.body }
+      rescue
+      end
+    end
+
+    worker1 = Thread.new do
+      begin
+        req1 = Net::HTTP::Get.new '/sleep40'
+        http1.start.request(req1) { |rep1| rep1.body }
+      rescue
+      end
+    end
+
+    old_pids = cluster.instance_variable_get(:@workers).map(&:pid)
+
+    start_time = Time.now.to_f
+    old_pids.each { |p| Process.kill :TERM, p }
+
+    # make sure four workers have booted
+    time = 0
+    until workers_booted >= 4 || time >= 30
+      sleep 2
+      time += 2
+    end
+
+    new_pids = cluster.instance_variable_get(:@workers).map(&:pid)
+
+    # should be empty if all old workers removed
+    old_waited = old_pids.map { |pid|
+      begin
+        Process.wait(pid, Process::WNOHANG)
+        pid
+      rescue Errno::ECHILD
+        nil # child is already terminated
+      end
+    }.compact
+
+    Thread.kill worker0
+    Thread.kill worker1
+    http0   = nil
+    http1   = nil
+    cluster = nil
+
+    l.stop
+    assert_kind_of Thread, t.join, "server didn't stop"
+
+    assert_operator (Time.now.to_f - start_time).round(2), :<, 32
```

Comment:
maybe add together what you're looking for explicitly here: 

```ruby
assert_operator (Time.now.to_f - start_time).round(2), :<, (conf.worker_shutdown_timeout + 30)
```

I forgot, where does the 30 come from?